examples/greeter-extension/README.md

```diff
@@ -0,0 +1,352 @@
+[Extension Point/extension pattern](https://wiki.eclipse.org/FAQ_What_are_extensions_and_extension_points%3F),
+which promotes loose coupling and offers great extensibility. There are many use
+cases in LoopBack 4 that fit into design pattern. For example:
+
+- `@loopback/boot` uses `BootStrapper` that delegates to `Booters` to handle
+  different types of artifacts
+- `@loopback/rest` uses `RequestBodyParser` that finds the corresponding
+  `BodyParsers` to parse request body encoded in different media types
+
+## Overview
+
+We'll use the following scenario to walk through important steps to organize the
+`greet` service that allows extensible languages - each of them being supported
+by a `Greeter` extension.
+
+![greeters](greeters.png)
+
+Various constructs from LoopBack 4, such as `Context`, `@inject.*`, and
+`Component` are used to build the service in an extensible fashion.
+
+## Define an extension point
+
+In our scenario, we want to allow other modules to extend or customize how
+people are greeted in different languages. To achieve that, we declare the
+`greeter` extension point, which declares a contract as TypeScript interfaces
+that extensions must conform to.
+
+### Define interface for extensions
+
+An extension point interacts with unknown number of extensions. It needs to
+define one or more interfaces as contracts that each extension must implement.
+
+```ts
+/**
+ * Typically an extension point defines an interface as the contract for
+ * extensions to implement
+ */
+export interface Greeter {
+  language: string;
+  greet(name: string): string;
+}
+```
+
+### Define class for the extension point
+
+Typically an extension point is defined as a TypeScript class and bound to a
+context. In our case, we mark `GreetingService` as the extension point that
+needs to access a list of greeters.
+
+```ts
+/**
+ * An extension point for greeters that can greet in different languages
+ */
+export class GreetingService {
+  constructor(
+    /**
+     * Inject a getter function to fetch greeters (bindings tagged with
+     * 'greeter')
+     */
+    @inject.getter(bindingTagFilter({extensionPoint: 'greeter'}))
+    private greeters: Getter<Greeter[]>,
+  ) {}
+  // ...
+}
+```
+
+To customize metadata such as `id` for the extension point, we can use
+`@extensionPoint` to decorate the class, such as:
+
+```ts
+@extensionPoint('greeter')
+export class GreetingService {}
+```
+
+#### Access extensions for a given extension point
+
+To simplify access to extensions for a given extension point, we use dependency
+injection to receive a `getter` function that gives us a list of greeters.
+
+```ts
+@extensionPoint('greeter')
+export class GreetingService {
+  constructor(
+    /**
+     * Inject a getter function to fetch greeters (bindings tagged with
+     * 'greeter')
+     */
+    @extensions() // Sugar for @inject.getter(filterByTag({extensionPoint: 'greeter'}))
+    private greeters: Getter<Greeter[]>, // ...
```

Comment:
Nouns are usually used for properties holding data. Functions/methods typically start with verbs.

```suggestion
    private getGreeters: Getter<Greeter[]>, // ...
```